### PR TITLE
introduce BoundingBox::is_neighbor()

### PR DESCRIPTION
**Pull request by jh66637**
I only want to know if bounding boxes are neighbors. If they are, get_neighbor_type() does too much work, since it checks which neighbor_type the bounding boxes are. 

During implementation, I found two issues:
* In my opinion ```side_length``` should always return positive values
* In test ```bounding_box_3``` the boxes C [0,2] and G [0.8,1.8] are considered no neighbors while they should be mergable neighbors. 

Given implementation fixes the issues.

### CODE REVIEW

File: source/base/bounding_box.cc [hunk 1]
```diff
@@ -55,42 +52,53 @@ BoundingBox<spacedim, Number>::merge_with(
                  other_bbox.boundary_points.first[i]);
       this->boundary_points.second[i] =
         std::max(this->boundary_points.second[i],
-                 other_bbox.boundary_points.second[i]);
+                    other_bbox.boundary_points.second[i]);
     }
```
Comment (bangerth) [inline source/base/bounding_box.cc:56]: This is a separate question if I understand you correctly, *independent* of the new function. Would you be willing to submit a separate pull request for this?

File: source/base/bounding_box.cc [hunk 2]
```diff
@@ -55,42 +52,53 @@ BoundingBox<spacedim, Number>::merge_with(
...
 
-      // Step 1: testing if the boxes are close enough to intersect
-      for (unsigned int d = 0; d < spacedim; ++d)
-        if (bbox1[0][d] * (1 - tolerance) > bbox2[1][d] ||
-            bbox2[0][d] * (1 - tolerance) > bbox1[1][d])
-          return NeighborType::not_neighbors;
-
       // The boxes intersect: we need to understand now how they intersect.
       // We begin by computing the intersection:
       std::array<double, spacedim> intersect_bbox_min;
```
Comment (bangerth) [inline source/base/bounding_box.cc:104]: Same here

File: source/base/bounding_box.cc [hunk 3]
```diff
@@ -55,42 +52,53 @@ BoundingBox<spacedim, Number>::merge_with(
...
+    {
+      // testing if the boxes are close enough to intersect
+      if ((other_bbox.boundary_points.second[i] <
+           this->boundary_points.first[i] - tolerance * side_length(i)) ||
+          (other_bbox.boundary_points.first[i] >
+           this->boundary_points.second[i] + tolerance * side_length(i)))
+        return false;
+    }
+  return true;
+}
```
Comment (bangerth) [inline source/base/bounding_box.cc:75]: This additional function is ok.

File: source/base/bounding_box.cc (hunk 1 above)
Comment (jh66637) [reply source/base/bounding_box.cc:56]: I am currently having problems with my clang-format integration. This is related to this issue. Sorry, I will remove this change.

File: source/base/bounding_box.cc (hunk 2 above)
Comment (jh66637) [reply source/base/bounding_box.cc:104]: So you want me to split the introduction of the function and the usage in the code? Is this because it automatically fixes a bug, or should I proceed this way in general?

File: source/base/bounding_box.cc (hunk 3 above)
Comment (simonsticko) [reply source/base/bounding_box.cc:75]: I would like to suggest a different name for this new function, e.g., `have_nonzero_intersection` or `intersection_is_nonzero` better describes what this function does.

Comment (jh66637) [reply source/base/bounding_box.cc:75]: If we change the name, I think we would also have to change the name of the corresponding ```NeighborType``` to stay consistent, which might break third party code. 
That said, I would prefer the proposed name as well.

File: source/base/bounding_box.cc [hunk 4]
```diff
@@ -215,7 +222,8 @@ BoundingBox<spacedim, Number>::side_length(const unsigned int direction) const
 {
   AssertIndexRange(direction, spacedim);
 
-  return boundary_points.second[direction] - boundary_points.first[direction];
+  return std::abs(boundary_points.second[direction] -
+                  boundary_points.first[direction]);
```
Comment (bangerth) [inline source/base/bounding_box.cc:226]: Isn't it an invariant of the class that the second point is always to the right of the first? I would much rather add an assertion to all of the places where one can set the boundary points to enforce the invariant, than to add the `abs` here.

File: tests/base/bounding_box_3.output [hunk 5]
```diff
@@ -65,7 +65,7 @@ DEAL::Is neighbor G with A: 3
 DEAL::Vice-versa: 3
 DEAL::Is neighbor G with B: 3
 DEAL::Vice-versa: 3
-DEAL::Is neighbor G with C: 0
+DEAL::Is neighbor G with C: 3
```
Comment (bangerth) [inline tests/base/bounding_box_3.output:68]: Why does the test output change?

**Review by bangerth — COMMENTED** [review]
Let's get back to this PR and get it merged! Here are just a couple of comments.

File: tests/base/bounding_box_3.output (hunk 5 above)
Comment (jh66637) [reply tests/base/bounding_box_3.output:68]: The boxes C [0,2] and G [0.8,1.8] are mergable neighbors (3). Currently, it is considered no neighbor (0).

File: source/base/bounding_box.cc (hunk 4 above)
Comment (jh66637) [reply source/base/bounding_box.cc:226]: I took it from the point_inside() function, but I thin you are right.

File: tests/base/bounding_box_3.output (hunk 5 above)
Comment (bangerth) [reply tests/base/bounding_box_3.output:68]: Ah, thanks!

File: tests/base/bounding_box_8.cc [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// test for BoundingBox<int dim> which tests the function
+// is_neighbor()
```
Comment (bangerth) [inline tests/base/bounding_box_8.cc:18]: need to update the function name here

**Review by bangerth — COMMENTED** [review]
OK if you address the one thing about the old function name in a comment and if you merge the various commits into one!

**Review by bangerth — COMMENTED** [review]
Also, feel free to put an entry into `doc/news/changes/minor` -- you should get credit in the changelog where credit is undoubtedly due!

**Comment by jh66637** [comment]
I applied requested changes :)

**Comment by bangerth** [comment]
Thank you! The contribution is much appreciated, and I hope that many more will follow! :-)

/rebuild

**Review by bangerth — APPROVED** [review]
